### DISCUSSION TITLE
Unreal crash reporter questions

### DISCUSSION THREAD

**Discussion by kristjanvalur** [getsentry/sentry-unreal#383] [Q&A]
I am unclear on the role of unreal's crash reporter in the context of sentry-unreal.  I've been using the sentry-native for a long time in my own plugin, and just catching crashes as they happen.  But stakeholders are interested in a more "visible" crash expecience for client end-users, rather than the application just popping out of existence.

1. If using native crash handling (u.e. 5.2 and above, by disabling ue's internal crash handlers), is there a role for the UnrealCrashHandler?  Do these two things go hand in hand?  If I have unreal-sentry enabled, already silently delivering crashes, is it possible to _also_ have the CrashHandler active in the client?  Or is this an either/or scenario?
2. With the crash handler, it is recommended to package "debug files" with the client.  By "debug files", does this mean debug symbols, i.e. .pdb files on windows?  Why is this needed when we are already uploading our symbol files to sentry during building?  We'd rather not have the symbol files in the hands of end users.

Does anyone have the answer to these questions?

**Comment by tustanivsky** [comment]
> If using native crash handling (u.e. 5.2 and above, by disabling ue's internal crash handlers), is there a role for the UnrealCrashHandler? Do these two things go hand in hand? If I have unreal-sentry enabled, already silently delivering crashes, is it possible to also have the CrashHandler active in the client? Or is this an either/or scenario?

Starting from UE 5.2 one can choose which crash capturing mechanism to use exclusively - `sentry-native` or `UnrealCrashHandler`. Both can't be active simultaneously since the `FPlatformMisc::SetCrashHandlingType` you've mentioned in #380 essentially disables one or another depending on its input. However, in older engine versions `UnrealCrashHandler` might be the only option available if talking about their binary distributions.

> With the crash handler, it is recommended to package "debug files" with the client. By "debug files", does this mean debug symbols, i.e. .pdb files on windows? Why is this needed when we are already uploading our symbol files to sentry during building? We'd rather not have the symbol files in the hands of end users.

If you're referencing [this guide](https://docs.sentry.io/platforms/unreal/configuration/setup-crashreporter/) here it primarily describes how to configure UE in order to make it work with Sentry without `sentry-unreal` plugin at all. I think including debug symbols is recommended mainly for user convenience so that they can be found more easily and uploaded to Sentry manually. Yet if Sentry plugin is enabled and it is configured to automatically handle symbol uploading then of course including debug symbols in build doesn't make much sense.

**Reply by kristjanvalur** [reply]
Thank you.  I think the documentation I was stumbling over is this:

> To get the most out of Sentry, crash reports are required to include debug information. In order for Sentry to be able to process the crash report and translate memory addresses to meaningful information like function names, module names and line numbers, the crash itself **must include debug** information and also, [symbols need to be uploaded to Sentry](https://docs.sentry.io/platforms/unreal/configuration/setup-crashreporter/#upload-debug-symbols).

this what is confusing to me.  Crash including debug information (presumably the toggle in "packaging", which now says "include debug symbols for shipping builds"), and also "symbols uploaded to sentry".

I think that the documentation might benefit from being updated, explaining how you can __either__ have automatic plugin integration and uploading of crashes __or__ depend on the unreal crash handler.


Here's a possible feature request:  The unreal crash handler has the advantage of being customer facing.  Would it be possible to use the CrashHandler in "client" builds (or another build target configuration), while sticking to automatic crash handling for, for example, server builds?

**Comment by tustanivsky** [comment]
> I think that the documentation might benefit from being updated, explaining how you can either have automatic plugin integration and uploading of crashes or depend on the unreal crash handler.

Noted! We'll make sure this topic is covered in the next documentation update.

> Here's a possible feature request: The unreal crash handler has the advantage of being customer facing. Would it be possible to use the CrashHandler in "client" builds (or another build target configuration), while sticking to automatic crash handling for, for example, server builds?

I believe you already can do that with the current implementation. In plugin settings `Misc -> Enable for Build Target Types` option gives you [control](https://github.com/getsentry/sentry-unreal/blob/1c60453b413820182d19429f3e526c527f0956ac/plugin-dev/Source/Sentry/Private/SentrySubsystem.cpp#L72C15-L72C15) over targets for which `sentry-native` has to be initialized and thus handle crashes automatically. At the same time `CrashHandler` should be capable of capturing crashes for all other targets that were disabled there.

The above example setup might be the following:

<img width="392" alt="image" src="https://github.com/getsentry/sentry-unreal/assets/3168564/8cc54855-5e7b-49b3-b198-f6edd9596dc4">

For `Server` target `sentry-native` gets initialized as usual and disables `CrashHandler` by calling `FPlatformMisc::SetCrashHandlingType(ECrashHandlingType::Disabled)`. Otherwise, the native integration init step is skipped and  `CrashHandler` kicks in instead and presents the user with whatever custom feedback upon crash is captured.

**Reply by kristjanvalur** [reply]
Thank you for your clarification.  I will run some experiments.

**Comment by RaiaN** [comment]
@kristjanvalur 
> But stakeholders are interested in a more "visible" crash expecience for client end-users, rather than the application just popping out of existence.
Are you talking about more visual-like debugging i.e. crash video / screenshot? I've implemented Crash video recording (last N seconds) which allow game developers to easily track what player did RIGHT BEFORE the crash
